File: answer.py
import importlib
import configparser
import load
from colorama import Fore
import base64
import os
from typing import List, Optional, Dict, Any, Tuple
import time
import requests
import json
import re
from chatContent import ChatContent

# ================= 配置加载（对齐 Answer.cs 构造函数） =================
config = configparser.ConfigParser()
config.read('config.ini', encoding='utf-8')


def _cfg(key: str) -> str:
    return config['general'][key]


modelName: str = _cfg('modelname')
serverUrl: str = _cfg('server_url')
isVisionModel: bool = _cfg('isvisionmodel').lower() == 'true'
maxImageCount: int = int(_cfg('maximagecount'))
remoteServerTimeout: int = int(_cfg('remote_server_timeout'))
forceOllamaAPI: bool = config.getboolean('general', 'forceollamaapi', fallback=False)
apiKey: str = _cfg('api_key')
# 系统提示从 config.ini 的 system 键读取（沿用旧版方式）
sysPmpt: str = config.get('general', 'system', fallback='')

useOllama = False
builtInLanguageModel = False
tinylm = None

# 后端选择（对齐 Answer.cs 构造函数）
if forceOllamaAPI:
    useOllama = True
    serverUrl+='/api/chat'
elif serverUrl.lower() == 'ollama':
    serverUrl = 'http://localhost:11434/api/chat'
    useOllama = True
elif serverUrl.lower() == 'builtin':
    builtInLanguageModel = True
# 否则 server_url 是用户自定义的 base URL（如 http://192.168.1.100:8000/v1）

MAX_LENGTH = 2048  # 对齐 C#；非流式请求，未用于截断

# extra.json 中的字段会合并进请求体并覆盖同名键（对齐 C#）
extra: Dict[str, Any] = {}
try:
    with open('extra.json', 'r', encoding='utf-8') as f:
        extra = json.loads(f.read())
except Exception:
    pass

# ================= 工具函数 =================

def _image_to_base64(path: str) -> str:
    with open(path, 'rb') as f:
        return base64.b64encode(f.read()).decode('utf-8')


def isTime(text: str) -> Tuple[bool, List[str]]:
    """对齐 Answer.cs 的 IsTime：校验小时 0-23、分钟 0-59"""
    pattern1 = r'\(?([0-2]?[0-9]):([0-5][0-9])\)?'
    pattern2 = r'\(?([0-2]?[0-9])\.([0-5][0-9])\)?'
    valid_times: List[str] = []
    for pattern in (pattern1, pattern2):
        for m in re.finditer(pattern, text):
            try:
                h = int(m.group(1))
                minute = int(m.group(2))
            except ValueError:
                continue
            if 0 <= h <= 23 and 0 <= minute <= 59:
                valid_times.append(f'{h:02d}:{minute:02d}')
    return len(valid_times) > 0, valid_times


def _concatenate_text(text_list: List[ChatContent], images: List[str]) -> List[Dict[str, Any]]:
    """对齐 Answer.cs 的 ConcatenateText。

    - 只收集同时出现在全局 images 列表（已按 MaxImageCount 截断）中的图片；
    - 只有用户消息携带图片，assistant 消息只带文本；
    - 既没有文本、也没有可发送的图片 → 整条消息跳过（纯图片消息必须保留）；
    - Ollama 用纯 base64 数组；OpenAI 兼容 API 用 data: URI 分段数组。
    """
    messages: List[Dict[str, Any]] = []

    for t in text_list:
        image_b64: List[str] = []        # 纯 base64 —— Ollama 使用
        image_data_urls: List[str] = []  # data: URI —— OpenAI 兼容 API 使用
        for img in t.imagePaths:
            if img not in images:
                continue
            b64 = _image_to_base64(img)
            mime = 'image/png' if img.lower().endswith('.png') else 'image/webp' if img.lower().endswith('.webp')  else 'image/jpeg'
            image_b64.append(b64)
            image_data_urls.append(f'data:{mime};base64,{b64}')

        has_text = bool(t.text)
        attach_images = (not t.ownByMyself) and len(image_b64) > 0

        if not has_text and not attach_images:
            continue

        text = str(t) if has_text else ''

        message: Dict[str, Any] = {
            'role': 'assistant' if t.ownByMyself else 'user',
        }

        if useOllama:
            # Ollama /api/chat 格式：images 是与 content 平级的【纯 base64】数组（不带 data: 前缀）
            message['content'] = text
            if attach_images:
                message['images'] = image_b64
        else:
            # OpenAI 兼容格式：content 可以是字符串，也可以是分段数组
            if not attach_images:
                message['content'] = text
            else:
                parts: List[Any] = [{'type': 'text', 'text': text}]
                for url in image_data_urls:
                    parts.append({'type': 'image_url', 'image_url': {'url': url}})
                message['content'] = parts

        messages.append(message)

    return messages


# ================= 核心推理函数 =================
def getAnswer(text: List[ChatContent], systemPrompt: str = 'auto') -> Tuple[Optional[str], int]:
    totalTokens = 0
    if text is None or len(text) == 0:
        return '', 0

    # 内置模型（对齐 Answer.cs Builtin 分支：从后往前找第一条非空、非自己发的消息）
    if builtInLanguageModel:
        global tinylm
        for t in reversed(text):
            if not t.text or t.ownByMyself:
                continue
            if tinylm is None:
                tinylm = importlib.import_module('TinyLangJaccard')
            return tinylm.answer(t.text), 0
        return '', 0

    # 系统提示（"auto" → config.ini 的 system；"" / "None" → 空；否则用传入值）
    if systemPrompt == 'auto':
        final_system_prompt = sysPmpt
    elif systemPrompt == '' or systemPrompt == 'None':
        final_system_prompt = ''
    else:
        final_system_prompt = systemPrompt

    # 收集图片：从后往前，跳过自己的消息，直到 MaxImageCount（对齐 C#）
    image_list: List[str] = []
    if isVisionModel:
        for t in reversed(text):
            if not t.ownByMyself:
                for img in t.imagePaths:
                    if os.path.exists(img):
                        image_list.append(img)
                        if len(image_list) >= maxImageCount:
                            break
                    else:
                        print(f'× 没有找到图片 {img}')
                if len(image_list) >= maxImageCount:
                    break

    # 构建 messages
    messages: List[Dict[str, Any]] = []
    if final_system_prompt:
        messages.append({'role': 'system', 'content': final_system_prompt})
    messages += _concatenate_text(text, image_list)

    # 构造请求体（对齐 C#）
    request_body: Dict[str, Any] = {
        'model': modelName,
        'messages': messages,
        'stream': False,
    }
    for k, v in extra.items():
        request_body[k] = v

    # 调试：写出完整请求体（对齐 C# dest.json）
    try:
        with open('dest.json', 'w', encoding='utf-8') as f:
            json.dump(request_body, f, ensure_ascii=False, indent=2)
    except Exception:
        pass

    # Headers（对齐 C#：API_KEY 非空且不是 localhost 时才带 Bearer）
    headers = {'Content-Type': 'application/json'}
    if apiKey and 'localhost' not in serverUrl and '127.0.0.1' not in serverUrl:
        headers['Authorization'] = f'Bearer {apiKey}'

    url = serverUrl if useOllama else f'{serverUrl}/chat/completions'

    start_time = time.time()
    try:
        print(f'Sending request to: {url}')

        resp = requests.post(url, json=request_body, headers=headers, timeout=remoteServerTimeout)
        response_body = resp.content.decode('utf-8', errors='replace')

        if resp.status_code < 200 or resp.status_code >= 300:
            print(f'API Error: {resp.status_code} - {response_body}')
            return None, 0

        print('\n\nResponse:\n')
        print(response_body)

        doc = json.loads(response_body)

        if useOllama:
            answer_text = doc.get('message', {}).get('content')
        else:
            answer_text = doc.get('choices', [{}])[0].get('message', {}).get('content')

        # Token 用量（对齐 C#：只读 prompt_tokens / completion_tokens / total_tokens，缺失为 0）
        usage = doc.get('usage')
        if usage is not None:
            prompt_tokens = usage.get('prompt_tokens', 0) or 0
            completion_tokens = usage.get('completion_tokens', 0) or 0
            total_tokens = usage.get('total_tokens', 0) or 0
            print(f'Token 用量: 输入 {prompt_tokens} | 输出 {completion_tokens} | 总计 {total_tokens}')
            totalTokens = total_tokens

        # 推理内容（Deepseek reasoning_content / Ollama thinking，对齐 C#）
        try:
            reason = doc.get('choices', [{}])[0].get('message', {}).get('reasoning_content')
            if reason is None:
                reason = doc.get('message', {}).get('thinking')
        except Exception:
            reason = None
        if reason is not None:
            print(f'{Fore.LIGHTBLACK_EX}<think>\n{reason}\n</think>{Fore.RESET}')

        elapsed = time.time() - start_time
        print(f'用时 {elapsed:.2f}s')

        if answer_text is not None:
            answer_text = answer_text.strip()
        print(answer_text if answer_text is not None else '')

        return answer_text, totalTokens

    except Exception as e:
        print(f'HTTP request failed: {e}')
        return None, 0


def get_answer_as_string(text: str, system_prompt):
    return getAnswer([ChatContent(username='', imagePaths=[], text=text, time='', ownByMyself=False)], system_prompt)


# ================= 测试入口 =================
if __name__ == '__main__':
    c = ChatContent(
        username='', imagePaths=[],
        text='解释图片\n2\n3\n4\n5',
        time=time.strftime('%Y-%m-%d %H:%M:%S', time.localtime()),
        ownByMyself=False,
    )
    c2 = ChatContent(
        username='', imagePaths=[r'D:\Pictures\111.PNG'],
        text='12345',
        time=time.strftime('%Y-%m-%d %H:%M:%S', time.localtime()),
        ownByMyself=True,
    )
    c3 = ChatContent(
        username='', imagePaths=[],
        text='678910',
        time=time.strftime('%Y-%m-%d %H:%M:%S', time.localtime()),
        ownByMyself=False,
    )

    # 测试 Ollama（对齐 C# Test()）
    useOllama = True
    serverUrl = 'http://localhost:8080/api/chat'
    modelName = 'qwen3.5:0.8b'

    print('\n=== Testing Ollama ===')
    answer = getAnswer([c, c2, c3])
    print(f'Answer: {answer}')

    # 测试 Built-in
    print('\n=== Testing Built-in Model ===')
    builtInLanguageModel = True
    answer = getAnswer([c, c2, c3])
    print(f'Answer: {answer}')

    # 测试 OpenAI Compatible
    print('\n=== Testing OpenAI Compatible API ===')
    builtInLanguageModel = False
    useOllama = False
    serverUrl = 'http://localhost:8000/v1'
    apiKey = '21r234242'
    answer = getAnswer([c, c2, c3])
    print(f'Answer: {answer}')

File: Answer.cs
using IniParser;
using IniParser.Model;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text;
using System.Text.Encodings.Web;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace QQPilot4
{
    internal class Answer
    {
        private readonly HttpClient _httpClient;
        private FileIniDataParser parser = new();
        private IniData? config;
        private string ModelName { get; set; } = "";
        private string ServerUrl { get; set; } = "";
        private bool IsVisionModel { get; set; }
        private int MaxImageCount { get; set; }
        private int RemoteServerTimeout { get; set; }
        private string ApiKey { get; set; } = "";
        private bool Builtin { get; set; } = false;
        private string SysPmpt { get; set; } = "";
        private TinyLangJaccardCS? TinyLangJaccard;
        private bool ForceOllamaAPI { get; set; }

        private
            readonly
            bool UseOllama=false;

        // 常量
        private const int MAX_LENGTH = 2048;

        public long TotalTokens
        {
            get; private set; }
        public Answer()
        {
            config = parser.ReadFile("config.ini", Encoding.UTF8);
            ModelName = config["general"]["modelname"];
            ServerUrl = config["general"]["server_url"];
            IsVisionModel = config["general"]["isvisionmodel"].Equals("true", StringComparison.OrdinalIgnoreCase);
            MaxImageCount = int.Parse(config["general"]["maximagecount"]);
            RemoteServerTimeout = int.Parse(config["general"]["remote_server_timeout"]);
            ForceOllamaAPI = bool.Parse(config["general"]["forceollamaapi"]);
            ApiKey = config["general"]["api_key"];
            SysPmpt = File.Exists("system.txt") ? File.ReadAllText("system.txt") : "";

            // 配置 HttpClient
            _httpClient = new HttpClient();
            _httpClient.Timeout = TimeSpan.FromSeconds(RemoteServerTimeout);
            if (ForceOllamaAPI)
            {
                UseOllama = true;
            }
            else if (ServerUrl.Equals("ollama", StringComparison.OrdinalIgnoreCase))
                {
                    ServerUrl = "http://localhost:11434/api/chat";
                    UseOllama = true;
                    //UseOllama = false;
            }
                else if (ServerUrl.Equals("builtin", StringComparison.OrdinalIgnoreCase))
                {
                    Builtin = true;
                }
            // 否则 ServerUrl 就是用户自定义的 base URL（如 http://192.168.1.100:8000/v1）
        }

        // --- 工具方法 ---

        private static string ImageToBase64(string path)
        {
            byte[] bytes = File.ReadAllBytes(path);
            return Convert.ToBase64String(bytes);
        }

        private static (bool hasTime, List<string> validTimes) IsTime(string text)
        {
            var pattern1 = @"\(?([0-2]?[0-9]):([0-5][0-9])\)?";
            var pattern2 = @"\(?([0-2]?[0-9])\.([0-5][0-9])\)?";

            var matches1 = Regex.Matches(text, pattern1);
            var matches2 = Regex.Matches(text, pattern2);

            var allMatches = matches1.Cast<Match>().Concat(matches2.Cast<Match>());
            var validTimes = new List<string>();

            foreach (var m in allMatches)
            {
                if (int.TryParse(m.Groups[1].Value, out int h) &&
                    int.TryParse(m.Groups[2].Value, out int min) &&
                    h >= 0 && h <= 23 && min >= 0 && min <= 59)
                {
                    validTimes.Add($"{h:D2}:{min:D2}");
                }
            }

            return (validTimes.Count > 0, validTimes);
        }

        private List<Dictionary<string, object>> ConcatenateText(List<ChatContent> textList, List<string> images)
        {
            var messages = new List<Dictionary<string, object>>();

            foreach (var t in textList)
            {
                // 只收集同时出现在全局 images 列表（已按 MaxImageCount 截断）中的图片
                var imageB64 = new List<string>();       // 纯 base64 —— Ollama 使用
                var imageDataUrls = new List<string>();  // data: URI  —— OpenAI 兼容 API 使用
                foreach (var img in t.ImagePaths)
                {
                    if (images.IndexOf(img) < 0) continue;

                    string b64 = ImageToBase64(img);
                    string mime = img.ToLower().EndsWith(".png") ? "image/png" : "image/jpeg";
                    imageB64.Add(b64);
                    imageDataUrls.Add($"data:{mime};base64,{b64}");
                }

                // 只有用户消息携带图片；assistant 消息只带文本
                bool hasText = !string.IsNullOrEmpty(t.Text);
                bool attachImages = !t.OwnByMyself && imageB64.Count > 0;

                // 既没有文本、也没有可发送的图片 → 整条消息跳过。
                // 注意：纯图片消息（Text 为空但有图片）必须保留，否则图片永远不会发出去。
                if (!hasText && !attachImages) continue;

                string text = hasText ? $"{t}" : "";

                var message = new Dictionary<string, object>
                {
                    ["role"] = t.OwnByMyself ? "assistant" : "user",
                };

                if (UseOllama)
                {
                    // Ollama /api/chat 格式：images 是与 content 平级的【纯 base64】数组（不带 data: 前缀）
                    message["content"] = text;
                    if (attachImages)
                    {
                        message["images"] = imageB64.ToArray();
                    }
                }
                else
                {
                    // OpenAI 兼容格式：content 可以是字符串，也可以是分段数组
                    if (!attachImages)
                    {
                        message["content"] = text;
                    }
                    else
                    {
                        var parts = new List<object>
                        {
                            new Dictionary<string, object>
                            {
                                ["type"] = "text",
                                ["text"] = text
                            }
                        };
                        foreach (var url in imageDataUrls)
                        {
                            parts.Add(new Dictionary<string, object>
                            {
                                ["type"] = "image_url",
                                ["image_url"] = new Dictionary<string, string>
                                {
                                    ["url"] = url
                                }
                            });
                        }
                        message["content"] = parts;
                    }
                }

                messages.Add(message);
            }

            return messages;
        }

        // --- 主逻辑：直接 POST 调用 API ---
        JsonSerializerOptions? jsonSerializerOptionsForPosting;
        JsonSerializerOptions? jsonSerializerOptionsForPrinting;
        public async Task<string?> GetAnswerAsync(List<ChatContent> text, string systemPrompt = "auto")
        {

            if (text == null || text.Count == 0) return "";

            // 内置模型
            if (Builtin)
            {
                foreach (var t in text.AsEnumerable().Reverse())
                {
                    if (string.IsNullOrEmpty(t.Text) || t.OwnByMyself) continue;
                    TinyLangJaccard ??= new TinyLangJaccardCS("datasetTiny.json");
                    return TinyLangJaccard.Answer(t.Text);
                }
                return "";
            }
            string extra = "{}";
            try
            {

                extra=File.ReadAllText("extra.json", new UTF8Encoding(false));
            }
            catch
            {

            }
           JsonDocument? extraJson;
            try
            {
             extraJson = JsonSerializer.Deserialize<JsonDocument>(extra);

            }
            catch
            {
                  extraJson=JsonSerializer.Deserialize<JsonDocument>("{}");
            }
            var mergedData = new Dictionary<string, object>();
            if (extraJson != null)
            {
                
                foreach (JsonProperty property in extraJson.RootElement.EnumerateObject())
                {
                    // property.Name 是键
                    // property.Value 是 JsonElement

                    // 注意：JsonElement 需要转换为具体类型才能存入 object
                    // 这里简单处理，根据类型转换
                    if (property.Value.ValueKind == JsonValueKind.String)
                        mergedData[property.Name] = property.Value.GetString()!;
                    else if (property.Value.ValueKind == JsonValueKind.Number)
                        mergedData[property.Name] = property.Value.GetDouble();
                    else if (property.Value.ValueKind == JsonValueKind.True || property.Value.ValueKind == JsonValueKind.False)
                        mergedData[property.Name] = property.Value.GetBoolean();
                    else if (property.Value.ValueKind == JsonValueKind.Null)
                        mergedData[property.Name] = "null";
                    else if (property.Value.ValueKind == JsonValueKind.Object)
                        // 嵌套对象可以递归处理，或者直接用 Utf8JsonWriter 写入
                        mergedData[property.Name] = JsonSerializer.Deserialize<object>(property.Value.GetRawText())!;
                    else if (property.Value.ValueKind == JsonValueKind.Array)
                        mergedData[property.Name] = JsonSerializer.Deserialize<object[]>(property.Value.GetRawText()!)!;
                }
            }
            string finalJson = JsonSerializer.Serialize(mergedData);
            Log.Print(finalJson);

            //foreach(var (k,v) in extraJson )
            // 系统提示
            string finalSystemPrompt = systemPrompt switch
            {
                "auto" => SysPmpt,
                "" or "None" => "",
                _ => systemPrompt
            };

            // 收集图片
            var imageList = new List<string>();
            var reversedText = new List<ChatContent>(text.ToArray());
            reversedText.Reverse();
            foreach (var t in reversedText)
            {
                if (!t.OwnByMyself)
                {
                    foreach (var img in t.ImagePaths)
                    {
                        if (File.Exists(img))
                        {
                            imageList.Add(img);
                            if (imageList.Count >= MaxImageCount) break;
                        }
                        else
                        {
                            Log.Print($"× 没有找到图片 {img}",Log.Stat.WARN);
                        }
                    }
                    if (imageList.Count >= MaxImageCount) break;
                }
            }

            // 构建 messages
            var messages = new List<Dictionary<string, object>>();
            if (!string.IsNullOrEmpty(finalSystemPrompt))
            {
                messages.Add(new Dictionary<string, object>
                {
                    ["role"] = "system",
                    ["content"] = finalSystemPrompt
                });
            }

            messages.AddRange(ConcatenateText(text, imageList));

            // 构造请求体
            var requestBody = new Dictionary<string, object>
            {
                ["model"] = ModelName,
                ["messages"] = messages,
                ["stream"] = false,
            };

            foreach(KeyValuePair<string, object> k in mergedData)
            {
                requestBody[k.Key]= k.Value;
            }

            jsonSerializerOptionsForPosting ??= new JsonSerializerOptions { WriteIndented = false };
            string json = JsonSerializer.Serialize(requestBody,jsonSerializerOptionsForPosting! );


            var content = new StringContent(json, Encoding.UTF8, "application/json");


            // 设置 Headers
            if (!string.IsNullOrEmpty(ApiKey) && !ServerUrl.Contains("localhost") && !ServerUrl.Contains("127.0.0.1"))
            {
                _httpClient.DefaultRequestHeaders.Authorization =
                    new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", ApiKey);
            }
            else
            {
                _httpClient.DefaultRequestHeaders.Authorization = null;
            }   

            try
            {
                var startTime = DateTime.UtcNow;
                HttpResponseMessage response;
                jsonSerializerOptionsForPrinting ??= new JsonSerializerOptions
                {
                    Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping,
                    WriteIndented = true // 可选：美化输出
                };
                try
                {

                File.WriteAllText("dest.json", JsonSerializer.Serialize(requestBody, jsonSerializerOptionsForPrinting!));
                }catch
                {

                }

                if (! UseOllama)
                {

           
                    Log.Print($"Sending request to: {ServerUrl}/chat/completions");


                    response = await _httpClient.PostAsync($"{ServerUrl}/chat/completions", content);
                }
                else
                {
                    Log.Print($"Sending request to: {ServerUrl}");

                    response = await _httpClient.PostAsync(ServerUrl, content);

                }
                string responseBody = await response.Content.ReadAsStringAsync();

                if (!response.IsSuccessStatusCode)
                {
                    Log.Print($"API Error: {response.StatusCode} - {responseBody}",Log.Stat.ERROR);
                    return null;
                }
                

                Log.Print($"\n\nResponse:\n");
                Log.Print(responseBody.ToString());

                using JsonDocument doc = JsonDocument.Parse(responseBody);
                string? answer = null;
                if (UseOllama)
                {
                    answer = doc.RootElement
                        .GetProperty("message")
                        .GetProperty("content")
                        .GetString();
                }
                else
                {
                    answer = doc.RootElement
                        .GetProperty("choices")[0]
                        .GetProperty("message")
                        .GetProperty("content")
                        .GetString();
                }
                try
                {
                    if (doc.RootElement.TryGetProperty("usage", out JsonElement usage))
                    {
                        int promptTokens = usage.TryGetProperty("prompt_tokens", out var pt) ? pt.GetInt32() : 0;
                        int completionTokens = usage.TryGetProperty("completion_tokens", out var ct) ? ct.GetInt32() : 0;
                        int totalTokens = usage.TryGetProperty("total_tokens", out var tt) ? tt.GetInt32() : 0;

                        Log.Print($"Token 用量: 输入 {promptTokens} | 输出 {completionTokens} | 总计 {totalTokens}");
                        TotalTokens += totalTokens;
                    }
                    string? reason = doc.RootElement //Deepseek
                        .GetProperty("choices")[0]
                        .GetProperty("message")
                        .GetProperty("reasoning_content")
                        .GetString();  
                    reason ??= doc.RootElement  //Ollama
                        .GetProperty("message")
                        .GetProperty("thinking")
                        .GetString();

                    if (reason is not null)
                    {
                        Log.SetColor(ConsoleColor.Gray);
                        Log.Print($"<think>\n{reason}\n</think>");
                        
                    }
                } catch (Exception ex)
                {
                    //Log.Print(ex.ToString(),Log.Stat.ERROR);
                }

                var elapsed = (DateTime.UtcNow - startTime).TotalSeconds;
                Log.Print($"用时 {elapsed:F2}s");
                Log.Print(answer?.Trim()??"");

                return answer?.Trim();
            }
            catch (Exception ex)
            {
                Log.Print($"HTTP request failed: {ex.Message}",Log.Stat.ERROR);
                return null;
            }
        }

        // 同步版本
        public string? GetAnswer(List<ChatContent> text, string systemPrompt = "auto")
        {
            return GetAnswerAsync(text, systemPrompt).GetAwaiter().GetResult();
        }

        public void Test()
        {
            try
            {
                //UseOllama = true;
                ServerUrl = "http://localhost:8080/api/chat";
                ForceOllamaAPI= true;
                //ChatContent c = ;
                Log.Print($"[ASSISTANT]: {GetAnswer([
                    new("Username1", [], "你好", DateTime.Now.ToShortDateString(), false),
                    //new("Username3", [], "1",DateTime.Now.ToShortDateString(), false),
                    //new("Username2", [], "1", DateTime.Now.ToShortDateString(), false),
                    //new("Username4", [], "1", DateTime.Now.ToShortDateString(), false),
                    //new("Username5", ["C:\\Users\\Develop\\Downloads\\juniorcuisine-template-26.2\\src\\main\\resources\\assets\\juniorcuisine\\textures\\item\\slime_podding.png"], "这是什么图片 /no_think", "", false),
                    //new("Username5", ["C:\\Users\\Develop\\Downloads\\juniorcuisine-template-26.2\\src\\main\\resources\\assets\\juniorcuisine\\textures\\item\\magma_cream_podding.png"], "这是什么图片 /no_think", "", false),
                    //new("Username5", ["C:\\Users\\Develop\\Downloads\\juniorcuisine-template-26.2\\src\\main\\resources\\assets\\juniorcuisine\\textures\\item\\nether_stew.png"], "这是什么图片 /no_think", "", false),
                    ])}");
            }
            catch (Exception ex)
            {
                Log.Print($"Test failed: {ex.Message}"  , Log.Stat.ERROR);
            }
        }
    }
}